ntp-daemon/src/system.rs

```diff
@@ -0,0 +1,184 @@
+    mut reset_epoch: ResetEpoch,
+    global_system_snapshot: Arc<tokio::sync::RwLock<SystemSnapshot>>,
+    mut msg_for_system_rx: mpsc::Receiver<MsgForSystem>,
+    reset_tx: watch::Sender<ResetEpoch>,
+) -> Result<(), Box<dyn Error>> {
+    let mut controller = ClockController::new(UnixNtpClock::new());
+    let mut snapshots = Vec::with_capacity(peers.len());
+
+    while let Some(msg_for_system) = msg_for_system_rx.recv().await {
+        peers.receive_update(msg_for_system, reset_epoch);
+
+        // remove snapshots from previous iteration
+        snapshots.clear();
+
+        // add all valid measurements to our list of snapshots
+        snapshots.extend(peers.valid_snapshots());
+
+        let ntp_instant = NtpInstant::now();
+        let system_poll = global_system_snapshot.read().await.poll_interval;
+        let result = FilterAndCombine::run(config, &snapshots, ntp_instant, system_poll);
+
+        let clock_select = match result {
+            Some(clock_select) => clock_select,
+            None => {
+                info!("filter and combine did not produce a result");
+                continue;
+            }
+        };
+
+        let offset_ms = clock_select.system_offset.to_seconds() * 1000.0;
+        let jitter_ms = clock_select.system_jitter.to_seconds() * 1000.0;
+        info!(offset_ms, jitter_ms, "system offset and jitter");
+
+        let adjust_type = controller.update(
+            clock_select.system_offset,
+            clock_select.system_jitter,
+            clock_select.system_peer_snapshot.root_delay,
+            clock_select.system_peer_snapshot.root_dispersion,
+            clock_select.system_peer_snapshot.leap_indicator,
+            clock_select.system_peer_snapshot.time,
+        );
+
+        // Handle situations needing extra processing
+        match adjust_type {
+            ClockUpdateResult::Panic => {
+                panic!(
+                    r"Unusually large clock step suggested,
+                            please manually verify system clock and reference clock 
+                                 state and restart if appropriate."
+                )
+            }
+            ClockUpdateResult::Step => {
+                peers.reset_all();
+
+                reset_epoch = reset_epoch.inc();
+                reset_tx.send_replace(reset_epoch);
+            }
+            _ => {}
+        }
+
+        // Handle updating system snapshot
+        if let ClockUpdateResult::Ignore = adjust_type {
+            // ignore this update
+        } else {
+            let mut global = global_system_snapshot.write().await;
+            global.poll_interval = controller.preferred_poll_interval();
+            global.leap_indicator = clock_select.system_peer_snapshot.leap_indicator;
+        }
+    }
+
+    // the channel closed and has no more messages in it
+    Ok(())
+}
+
+#[derive(Debug, Clone, Copy)]
+enum PeerStatus {
+    Demobilized,
+    AwaitingReset,
+    Valid(PeerSnapshot),
+}
```

Comment:
Again, missing status for when the peer is just not acceptable for synchronization.

Reply:
maybe I'm something here, but this is not where the system applies filtering. This is just whatever the peers report. The `snapshots` contain valid peer snapshots (so any peer that has a measurement taken in the current reset epoch), and only then do we actually inspect the content to see if we should use this peer's snapshot. 

The initial state of a peer is the `AwaitingReset`, which really means that the system is waiting for a valid message from this peer (valid meaning from withing the current reset epoch).